Step3 - 지하철 구간 관계

seonje1 opened the pull request:
안녕하세요 3단계 미션 PR 입니다

질문이 있습니다.

현재 given절에서 데이터 세팅할때 실제 사용자 시나리오와 같이 API호출로 데이터를 세팅하고 있습니다.
실무에서 큰 도메인을 맡은 경우 이보다 객체그래프가 더 깊어질수도 있을것 같은데 이런 경우에도 API를 호출하여 데이터를 모두 세팅하는지, 아니면 다른 방식을 선택할수도있는지 질문드립니다!

감사합니다.

seonje1 committed:
구간 등록, 삭제 인수 테스트 시나리오 작성

seonje1 committed:
기능 구현하여 작성한 인수테스트 시나리오 통과

seonje1 committed:
구간등록 인수 테스트 시나리오 리팩토링

- truncate.sql사용 -> 동적으로 엔티티 조회하여 테이블 truncate하는 방식으로 변경
- sectionApiRequester생성하여 api호출 부분 분리

junwoochoi commented on src/main/java/subway/section/Section.java:
```diff
@@ -0,0 +1,59 @@
+
+    private int distance;
+
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "up_station_id", foreignKey = @ForeignKey(ConstraintMode.NO_CONSTRAINT))
+    private Station upStation;
+
+    @OneToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "down_station_id", foreignKey = @ForeignKey(ConstraintMode.NO_CONSTRAINT))
+    private Station downStation;
```
추후에 환승역이 가능해지면 OneToOne구조로 보기는 어려울 수 있습니다! 현재 요구사항에서는 문제없어서 다음주차 미션에서 한번 확인하고 넘어가주세요~

junwoochoi commented on src/main/java/subway/section/SectionControllerAdvice.java:
```diff
@@ -0,0 +1,18 @@
+    @ExceptionHandler(SectionException.class)
+    @ResponseStatus(HttpStatus.BAD_REQUEST)
+    public String test(SectionException e) {
+        e.printStackTrace();
```
e.printStackTrace()는 일반적으로 안티패턴으로 알려져있습니다. logger를 만들어서 log를 남겨보시는 것은 어떨까요?

junwoochoi commented on src/main/java/subway/section/SectionException.java:
```diff
@@ -0,0 +1,19 @@
+package subway.section;
+
+public class SectionException extends RuntimeException {
```
커스텀 예외 생성 👍

junwoochoi commented on src/main/java/subway/section/SectionRepository.java:
```diff
@@ -0,0 +1,18 @@
+            "and s.upStation.id = :stationId " +
+            "or s.downStation.id = :stationId"
+    )
+    List<Section> findDeleteSections(List<Section> sections, Long stationId);
```
사용하지 않는 메소드는 제거해주세요!

junwoochoi commented on src/test/java/subway/section/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,164 @@
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+public class SectionAcceptanceTest {
+
+    @Autowired
+    private TableTruncate tableTruncate;
```
각 인수테스트 클래스에서 반복되는 TableTruncate과 @SpringBootTest를 리팩토링해볼수있을까요?

junwoochoi commented on src/test/java/subway/TableTruncate.java:
```diff
@@ -0,0 +1,44 @@
+                .collect(Collectors.toList());
+    }
+
+    @Transactional
+    public void truncate() {
+        em.flush();
+        em.clear();
+        for (String tableName : tableNames) {
+            em.createNativeQuery(String.format(TRUNCATE_FORMAT, tableName)).executeUpdate();
+            em.createNativeQuery(String.format(ALTER_TABLE_FORMAT, tableName)).executeUpdate();
+        }
+    }
```
피드백 적용 👍 👍

junwoochoi commented on src/test/java/subway/section/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,164 @@
+
+    @BeforeEach
+    void setUp() {
+        tableTruncate.truncate();
+        stationApiRequester.createStationApiCall("잠실역");
+        stationApiRequester.createStationApiCall("용산역");
+        stationApiRequester.createStationApiCall("건대입구역");
+        stationApiRequester.createStationApiCall("성수역");
+
+        LineCreateRequest request = new LineCreateRequest("2호선", "green", 1L, 2L, 10);
+        lineApiRequester.createLineApiCall(request);
+    }
```
테스트 픽스처 생성 💯

junwoochoi commented on src/test/java/subway/section/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,164 @@
+    @Test
+    void generateSection() {
+        //when
+        SectionCreateRequest request = new SectionCreateRequest(2L, 3L, 5);
```
다른 개발자나 본인이 시간이 지나고 다시 코드를 보았을때 2L, 3L가 각각 어떤 역을 뜻하는지 바로 알기 쉬운 방향은 없을까요?
@BeforeEach에서 생성한 테스트 픽스쳐를 필드변수에 저장해보고 해당 변수를 활용해보면 어떨까요?
2L, 3L이 어떤 역을 뜻하는 지 안다는 것은 블랙박스 테스트 영역의 방향과도 멀어지는 것 같아요

junwoochoi commented on src/test/java/subway/section/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,164 @@
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.CREATED.value());
+
+        ExtractableResponse<Response> findLine = lineApiRequester.findLineApiCall(1L);
+        assertThat(getStationIds(findLine)).containsExactly(1L, 2L, 3L);
```
여기서도 1L, 2L, 3L보다는 생성하고 받아온 ID를 활용해보면 좋겠습니다~

junwoochoi commented on src/test/java/subway/section/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,164 @@
+        assertThat(getStationIds(findLine)).containsExactly(1L, 2L, 3L);
+    }
+
+    /**
+     * When 등록할 구간의 상행역이 노선에 등록되어있는 하행종점역이 아닌 구간을 등록하면
+     * Then 예외가 발생한다
+     */
+    @DisplayName("등록할 구간의 상행역이 노선에 등록되어있는 하행종점역이 아닌 경우 구간을 등록할 수 없다")
+    @Test
+    void generateSectionException() {
+        //when
+        SectionCreateRequest request = new SectionCreateRequest(3L, 4L, 5);
+
+        ExtractableResponse<Response> response = sectionApiRequester.generateSection(request, 1L);
+
+        //then
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value());
+        assertThat(response.asPrettyString()).isEqualTo("등록할 구간의 상행역이 노선에 등록되어있는 하행종점역이 아닌 경우 구간 등록이 불가능합니다.");
+    }
+
+    /**
+     * When 등록할 구간의 하행역이 이미 해당 노선에 등록되어있으면
+     * Then 예외가 발생한다
+     */
```
예외케이스 시나리오 검증 👍

junwoochoi commented on src/test/java/subway/section/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,164 @@
+        assertThat(response.asPrettyString()).isEqualTo("구간이 1개인 노선의 구간은 삭제할 수 없습니다.");
+    }
+
+    private static List<Long> getStationIds(ExtractableResponse<Response> findLine) {
```
static 메소드가 아니어도 괜찮을 것 같습니다!

junwoochoi commented on src/test/java/subway/section/SectionApiRequester.java:
```diff
@@ -0,0 +1,32 @@
+
+import static io.restassured.RestAssured.given;
+
+public class SectionApiRequester {
```
반복되는 요청 분리 💯

junwoochoi commented on src/main/java/subway/station/StationDataService.java:
```diff
@@ -0,0 +1,26 @@
+                .orElseThrow(() -> new LineException("존재하지 않는 역입니다."));
+    }
+
+    public StationResponse mappingToStationResponse(Station station) {
```
Station 객체 -> StationResponse객체로 맵핑하는 역할이 StationDataService의 책임이 맞을까요? 
서비스 단에서보다는 DTO단인 StationResponse내부 생성자에서 처리해봐도 좋을 것 같습니다!

mappingToLineResponse도 같은 의견입니다

junwoochoi commented on src/main/java/subway/line/Line.java:
```diff
@@ -25,6 +31,8 @@ public class Line {
     @JoinColumn(name = "down_station_id", foreignKey = @ForeignKey(ConstraintMode.NO_CONSTRAINT))
     private Station downStation;
 
+    @OneToMany(mappedBy = "line")
```
cascade, orphanRemoval옵션을 활용해보면 Line을 통해서만 section을 저장,제거,수정할 수 있는데 요런 방안은 어떻게 생각하시나요?

junwoochoi replied:
추가로 Line에 기존 단계에서 사용하던 upStation , downStation은 제거하고, sections를 활용하도록 기존 구간생성 로직도 수정 부탁드립니다~

junwoochoi commented on src/main/java/subway/line/Line.java:
```diff
@@ -60,8 +68,76 @@ public Station getDownStation() {
         this.color = color;
     }
+
+    public void generateSection(Section section) {
+
+        verifyDownStation(section);
+
+        verifyAlreadyStation(section);
+
+        this.downStation = section.getDownStation();
+        this.distance = this.distance + section.getDistance();
+        sections.add(section);
+    }
+
+    private void verifyDownStation(Section section) {
+        boolean isNotLineDownStation = !this.downStation.getId().equals(section.getUpStation().getId());
+
+        if (isNotLineDownStation) {
+            throw new LineException("등록할 구간의 상행역이 노선에 등록되어있는 하행종점역이 아닌 경우 구간 등록이 불가능합니다.");
+        }
```
Line내부에 List<Section>을 위한 메소드들이 많아졌는데, 일급컬렉션을 활용해서 구간관련 로직을 캡슐화해보면 어떨까요?

junwoochoi reviewed (changes requested):
안녕하세요! 요구사항에 맞추어 구현해주신 부분 잘 확인했습니다.
관련해서 몇가지 코멘트 달아보았는데 확인 부탁드립니다~!

추가로 질문주신 테스트 픽스처 생성에 관해서는 정확히 무조건 API로만 해야된다. 이런 정답이 있는 부분은 아닙니다. 현재는 강의 미션 진행중이고 API가 이미 존재하는 서비스이기 때문에 블랙박스 테스트에 맞게 기존 시나리오를 활용해서 사전데이터를 생성하시길 추천드리고 있습니다. 실무에서는 항상 데이터 생성을 위한 API가 존재하는 것은 아니기 때문에 각각의 상황에 맞게 EntityManger나 Repository를 활용하는 경우도 있고, @Sql을 활용하기도 합니다. 
어떻게 테스트 데이터를 생성하느냐보다는 어떻게 테스트 데이터를 생성하는 부분을 깔끔히 관리하느냐가 더 어렵더라구요 😅

seonje1 committed:
Exception 리팩토링

- 사용하지 않는 SectionException, SectionControllerAdvice 삭제
- LineControllerAdvice e.printStackTrace() -> log로 변경

seonje1 committed:
사용하지않는 메소드 삭제

seonje1 committed:
static 메소드 제거

seonje1 committed:
DTO 매핑 리팩토링

- 기존 Service계층에 존재하던 DTO 매핑 메서드를 DTO 내부에 정적팩토리메소드로 변경(LineResponse, StationResponse)

seonje1 committed:
엔티티 구조 변경

- Line엔티티에 sections 일급 컬렉션으로 변경
- Casecade, orphanRemoval를 이용하여 Line에서 Section을 저장, 제거, 수정 하도록 변경
- Line엔티티 상행종점역, 하행종점역, 거리 컬럼 제거 (sections 활용)

seonje1 committed:
인수테스트 시나리오 변수 네이밍 변경 - 의미를 알기 힘든 변수명들 한글로 수정

seonje1 committed:
테스트 코드 리팩토링

- 반복되어 사용되는 로직(@SpringBootTest, truncate 메소드 실행 로직) 리팩토링
- JUnit Extension을 이용하여 truncate 중복 제거
- @AcceptanceTest 어노테이션으로 @SpringBootTest와 @ExtentionWith 어노테이션 중복 제거

seonje1 committed:
Sections 필요없는 생성자 삭제, Line 메소드 순서 변경

junwoochoi commented on src/main/java/subway/line/entity/Line.java:
```diff
@@ -0,0 +1,59 @@
+
+    private String color;
+
+    @Embedded
+    private Sections sections = new Sections();
```
👍

junwoochoi commented on src/test/java/subway/section/SectionAcceptanceTest.java:
```diff
@@ -21,18 +19,15 @@
 
 @DisplayName("지하철노선 구간 관련 기능")
-@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+@AcceptanceTest
```
💯

junwoochoi reviewed (changes requested):
안녕하세요 선제님~!
피드백 적용 잘해주셨습니다! @AcceptanceTest를 만든 부분과 beforeeachcallback을 활용하신 부분도 인상깊었어요!
아래 이미지의 hidden으로 숨겨진 코멘트 부분은 확인을 놓치신 것 같은데 요 부분에 있는 1L, 2L 관련된  코멘트 부분만 한번 수정해서 요청 주시면 감사하겠습니다! 😄 
<img width="759" alt="image" src="https://github.com/next-step/atdd-subway-map/assets/26688947/3033df42-3820-412b-a218-2ddb471764d9">

seonje1 committed:
테스트 코드 리팩토링

- 변수명을 한글로 하여 읽기 쉽도록 수정

seonje1 committed:
ApiRequester의 메소드를 static으로 변경

junwoochoi commented on src/test/java/subway/line/LineAcceptanceTest.java:
```diff
@@ -23,12 +21,17 @@ public class LineAcceptanceTest {
     private final StationApiRequester stationApiRequester = new StationApiRequester();
     private final LineApiRequester lineApiRequester = new LineApiRequester();
 
+    Long 잠실역id;
+    Long 용산역id;
+    Long 건대입구역id;
+    Long 성수역id;
```
💯

junwoochoi reviewed (approved):
안녕하세요 선제님!
1주차 마지막 미션 진행하시느라 수고 많으셨습니다!
2주차 미션도 계속해서 화이팅 해주세요~ 🏅
